go-controller/pkg/ovn/controller/services/node_tracker.go

```diff
@@ -0,0 +1,210 @@
+			node, ok := obj.(*v1.Node)
+			if !ok {
+				return
+			}
+			nt.updateNode(node)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldObj, ok := old.(*v1.Node)
+			if !ok {
+				return
+			}
+			newObj, ok := new.(*v1.Node)
+			if !ok {
+				return
+			}
+			// Make sure object was actually changed and not pending deletion
+			if oldObj.GetResourceVersion() == newObj.GetResourceVersion() || !newObj.GetDeletionTimestamp().IsZero() {
+				return
+			}
+
+			nt.updateNode(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			node, ok := obj.(*v1.Node)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					klog.Errorf("couldn't understand non-tombstone object")
+					return
+				}
+				node, ok = tombstone.Obj.(*v1.Node)
+				if !ok {
+					klog.Errorf("couldn't understand tombstone object")
+					return
+				}
+			}
+			nt.removeNode(node.Name)
+		},
+	})
+
+	return nt
+
+}
+
+// updateNodeInfo updates the node info cache, and syncs all services
+// if it changed.
+func (nt *nodeTracker) updateNodeInfo(nodeName, switchName, routerName string, nodeIPs []string, podSubnets []*net.IPNet) {
+	ni := nodeInfo{
+		name:              nodeName,
+		nodeIPs:           nodeIPs,
+		podSubnets:        make([]net.IPNet, 0, len(podSubnets)),
+		gatewayRouterName: routerName,
+		switchName:        switchName,
+	}
+	for i := range podSubnets {
+		ni.podSubnets = append(ni.podSubnets, *podSubnets[i]) // de-pointer
+	}
+
+	nt.Lock()
+	if existing, ok := nt.nodes[nodeName]; ok {
+		if reflect.DeepEqual(existing, ni) {
+			nt.Unlock()
+			return
+		}
+	}
+
+	nt.nodes[nodeName] = ni
+	nt.Unlock()
+
+	klog.Infof("Node %s switch + router changed, syncing services", nodeName)
+	// Resync all services
+	nt.resyncFn()
+}
+
+// RemoveNode removes a node from the LB -> node mapper
+// We don't need to re-sync here, because any stale LBs
+// will eventually be cleaned up, and they don't have any cost.
+func (nt *nodeTracker) removeNode(nodeName string) {
+	nt.Lock()
+	defer nt.Unlock()
+
+	delete(nt.nodes, nodeName)
+}
+
+// UpdateNode is called when a node's gateway router / switch / IPs have changed
+// The switch exists when the HostSubnet annotation is set.
+// The gateway router will exist sometime after the L3Gateway annotation is set.
+func (nt *nodeTracker) updateNode(node *v1.Node) {
+	klog.V(2).Infof("Processing possible switch / router updates for node %s", node.Name)
+	hsn, err := util.ParseNodeHostSubnetAnnotation(node)
+	if err != nil || hsn == nil {
+		klog.Infof("Node %s has invalid / no HostSubnet annotations: %v", node.Name, err)
+		nt.removeNode(node.Name)
+		return
+	}
+
+	switchName := node.Name
+	grName := ""
+	ips := []string{}
+
+	// if the node has a gateway config, it will soon have a gateway router
+	// so, set the router name
+	gwConf, err := util.ParseNodeL3GatewayAnnotation(node)
+	if err != nil || gwConf == nil {
+		klog.Infof("Node %s has invalid / no gateway config: %v", node.Name, err)
```

Comment:
Should this be a Warning instead?

Reply:
No, since nodes are in this state until they're initialized. Nothing to do here until the gateway controller wakes up and has coffee.

Reply:
(tweaked the wording a bit)